### PR TITLE
Run now directly by the runner, add cancelation

### PR DESCRIPTION
**Pull request by Maeeen**

### CODE REVIEW

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala [hunk 1]
```diff
@@ -14,17 +14,20 @@ import scala.concurrent.Future
...
 import com.olegych.scastie.api
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.TimeoutException
 import org.eclipse.lsp4j.jsonrpc.messages.CancelParams
+import java.net.URI
+import java.nio.file.Paths
 
 
 object BspClient {
-  case class BspClientRun(output: List[String], instrumentation: Option[List[api.Instrumentation]] = None)
```
Comment (rochala) [inline scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala:27]: I don't see a point in creating case class for one field. You can do a type alias `type BspClientRun = ProcessBuilder`

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala [hunk 2]
```diff
@@ -146,33 +147,34 @@ class BspClient(private val workingDir: Path,
         .map(r => r.getItems())
```
Comment (rochala) [inline scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala:147]: add `.asScala` here so you can use it's api in further calls such as
`if (classes.size() == 0)` can become `if (classes.isEmpty)`. Try to always convert it to Scala API if you're doing anything more than getting the values.

You shouldn't also throw Errors here, instead try to return Optional value / Either[Erorr, Something].

If you decide to do so with Option or Either, you can even do `classes.headOption.flatMap(_.headOption)`

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala [hunk 3]
```diff
@@ -146,33 +147,34 @@ class BspClient(private val workingDir: Path,
...
-          }))
+          bspServer.jvmRunEnvironment(new JvmRunEnvironmentParams(java.util.List.of(targetId)))
+            .asScala.map((mainClass, _))
         })
-        .map(result => {
-          if (result.getStatusCode() == StatusCode.ERROR) throw FailedRunError(s"Failed run: $result.")
-          if (result.getStatusCode() == StatusCode.CANCELLED) throw FailedRunError(s"Cancelled run: $result")
-          result
+        .map({
+          case (mainClass, runSettings) if runSettings.getItems().isEmpty() => throw BspClient.FailedRunError("Could not find run settings.")
```
Comment (rochala) [inline scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala:159]: The same goes here, refactor it to Either

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala [hunk 4]
```diff
@@ -146,33 +147,34 @@ class BspClient(private val workingDir: Path,
...
+            envVars.toSeq : _*
+          )
+
+          BspClientRun(process)
+        }}).recover { e =>
           logMessages.remove(s"$id-run")
-          runMessageEvent.remove(id)
           throw e
         }
       })
```
Comment (rochala) [inline scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala:180]: Overall, as I told you before in the future ( not scala one, future work ), try to refactor it to separate calls so we can have final build methods constructed with calls like:

```scala
def build(blabla):
  initialize(blabla)
    .flatMap(reload(blablab))
    .flatMap(getMainClass(blabla))
    .flatMap(etc etc)
```

You can even do for comprehension then.

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala [hunk 5]
```diff
@@ -58,45 +70,85 @@ class ScliRunner {
...
-            pb.copy(line = pb.line.map(_ + totalOffset + 1))))
-          case other => throw other
-        } 
+
+        }})
+      )
+      javaProcesses.put(snippetId, runProcess)
+
+      // Wait for process to exit gracefully (or being killed by timeout…)
+      val f = Future { runProcess.exitValue() }
```
Comment (rochala) [inline scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala:117]: Yeah, Scala doesn't provide a good way to timeout runs.

You may try to use https://doc.akka.io/docs/akka/current/futures.html#after as we have Akka as dependency anyway, and when we will be migrating away from it we will probably do to some other library providing such possibility eg. cats-effect / monix.

Check if akka's `after` can be used to implement this. If so change it, if not leave it for now as it is.

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala [hunk 6]
```diff
@@ -58,45 +70,85 @@ class ScliRunner {
...
   }
 
   def end: Unit = {
     bspClient.end
+    javaProcesses.values.foreach(_.destroy())
     process.map(_.destroy())
   }
 
+  // Java processes
+  private val javaProcesses = HashMap[SnippetId, Process]() // mutable HashMap
```
Comment (rochala) [inline scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala:150]: HashMap is not concurrent safe. And I think we are going to support multiple runtimes at the same time. Try to use `TrieMap` instead.

**Review by rochala — CHANGES_REQUESTED** [review]
Overall, great job. I love the idea with `InputStream` with callback to pass it to process. Also I see that you've managed to run processes with java.

Most comments are about throwing errors and refactoring build method so it has one responsibility. Try to refactor them to either.

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala (hunk 6 above)
Comment (Maeeen) [reply scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala:150]: Done!

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala (hunk 3 above)
Comment (Maeeen) [reply scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala]: Done… :)

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala (hunk 2 above)
Comment (Maeeen) [reply scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala]: Done hehe

File: scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala (hunk 1 above)
Comment (Maeeen) [reply scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala:27]: Done, again!

**Comment by Maeeen** [comment]
> Overall, great job. I love the idea with InputStream with callback to pass it to process.

In fact, I do not even need it. ProcessIO has a method to handle output via callbacks :1st_place_medal: 

> Also I see that you've managed to run processes with java.

Yes, required some effort but done!

> Most comments are about throwing errors and refactoring build method so it has one responsibility. Try to refactor them to either.

All of your suggestions have been taken into account by yours truly. Feel free to check the updated code.

**Review by rochala — APPROVED** [review]

**Pull request merged**